moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp

```diff
@@ -247,6 +248,38 @@ class PlanningSceneInterface::PlanningSceneInterfaceImpl
     planning_scene_diff_publisher_.publish(planning_scene);
   }
 
+  bool setCollisions(bool set_to_allow, const std::vector<std::string>& link_group_1,
+                     const std::vector<std::string>& link_group_2)
+  {
+    bool allowed_or_disallowed = set_to_allow;
+    moveit_msgs::GetPlanningScene::Request request;
+    moveit_msgs::GetPlanningScene::Response response;
+    std::vector<std::string> result;
+    request.components.components = request.components.ALLOWED_COLLISION_MATRIX;
+    if (!planning_scene_service_.call(request, response))
+    {
+      ROS_WARN_NAMED("planning_scene_interface",
+                     "Could not fetch allowed collision matrix");
+      return false;
+    }
+
+    moveit_msgs::PlanningScene ps = response.scene;
+    collision_detection::AllowedCollisionMatrix acm(response.scene.allowed_collision_matrix);
+    // TODO felixvd: Add option to include the child links in each link group?
+    for (const auto& b : link_group_1)
+    {
+      if (link_group_2.empty())
+        acm.setEntry(b, allowed_or_disallowed);
+      else
+        for (const auto& b2 : link_group_2)
+          acm.setEntry(b, b2, allowed_or_disallowed);
+    }
+
+    ps.is_diff = true;
+    acm.getMessage(ps.allowed_collision_matrix);
+    return applyPlanningScene(ps);
+  }
+
```

Comment:
This method invites a rather evil kind of racing condition because you pass the whole collision matrix via ROS services between nodes twice.
I am not in favor of having this implementation upstream.
I proposed to change the diff behavior of diff-PlanningScenes a year ago in https://github.com/ros-planning/moveit/issues/964 but nothing came out of it until now.
Do you think you could move your logic to the PlanningScene class and and implement a safe version as proposed in the issue?

Reply:
True. It's safe enough for our purposes, but not generally. I pushed the layer that we have in our system right now and added to the discussion in the other thread.